src/main/java/jakarta/validation/Validation.java

```diff
@@ -327,8 +329,8 @@ private static class GetValidationProviderListAction implements PrivilegedAction
 
 		//cache per classloader for an appropriate discovery
 		//keep them in a weak hash map to avoid memory leaks and allow proper hot redeployment
-		private final WeakHashMap<ClassLoader, SoftReference<List<ValidationProvider<?>>>> providersPerClassloader =
-				new WeakHashMap<>();
+		private final Map<ClassLoader, SoftReference<List<ValidationProvider<?>>>> providersPerClassloader =
+				Collections.synchronizedMap(new WeakHashMap<ClassLoader, SoftReference<List<ValidationProvider<?>>>>());
```

Comment:
Are you sure you're not doing something wrong in your code? Because write operations to this map should be extremely rare and if you have issues, I wonder if you are not doing something wrong.
That being said, I think the change makes sense.

@yrodiere could you have a look at this change too?

Reply:
@gsmet thank you for the comments
there are few issues related to WeakHashMap not being synchronized , I have checked with JDK team too 

https://github.com/eclipse-ee4j/metro-jax-ws/issues/61
https://github.com/spring-projects/spring-loaded/issues/194
https://www.adam-bien.com/roller/abien/entry/endless_loops_in_unsychronized_weakhashmap